Fix race condition in SerialTaskQueue

Dr15Jones opened the pull request:
#### PR description:
- Fix race condition when a task empties at the same time a new task is added to an empty task queue.
- Improved task waiting in FWCore/Concurrency tests
  - Replaced looping over tbb::task_group::wait calls with use of edm::FinalWaitingTask.
  - Added another case where REQUIRE needs to be synchronized

This is meant to try to address infrequent failures of the test where the test hangs forever.
#### PR validation:

The unit test runs.

resolves https://github.com/cms-sw/framework-team/issues/1467
resolves https://github.com/cms-sw/framework-team/issues/1468

Dr15Jones committed:
Use FinalWaitTask in FWCore/Concurrency unit tests

The FinalWaitTask is more robust than doing a while loop over the tbb_group::wait call.
Also added SAFE_REQUIRE to protect the call to REQUIRE in more places.

Dr15Jones committed:
Fixed English in the documentation

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones for master.

It involves the following packages:

- FWCore/Concurrency (**core**)


@Dr15Jones, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
The failure happened just like in the IBs with the previous code

```
-------------------------------------------------------------------------------
SerialTaskQueueChain
  stress test
```

This time it ran for 40 tries out of 100 before getting stuck.

Although I'd argue this PR is an improvement over the old code, it did not fix the underlying failure.

Dr15Jones committed:
Fix race condition in SerialTaskQueue

There was a race where if the queue was empty right when the running task finished and the task sets m_taskChosen and see that the queue is empty and right then a new task is added and that thread sees m_taskChosen was set, then neither threads will start the next task.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48582/45549

cmsbuild commented:
Pull request #48582 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

Dr15Jones commented:
please test
